internal/alb/tags/tags.go

```diff
@@ -0,0 +1,161 @@
+	Arn string
+
+	// Tags is a map[key]value of tags for the resource
+	Tags map[string]string
+}
+
+// NewTags allocates an empty Tags struct
+func NewTags(m ...map[string]string) *Tags {
+	t := &Tags{Tags: make(map[string]string)}
+	for i := range m {
+		for k, v := range m[i] {
+			t.Tags[k] = v
+		}
+	}
+	return t
+}
+
+func (t *Tags) Copy() *Tags {
+	n := NewTags()
+	for k, v := range t.Tags {
+		n.Tags[k] = v
+	}
+	return n
+}
+
+// TagsController manages tags on a resource
+type TagsController interface {
+	Reconcile(context.Context, *Tags) error
+}
+
+// NewTagsController constructs a new tags controller
+func NewTagsController(ec2 ec2iface.EC2API, elbv2 elbv2iface.ELBV2API) TagsController {
+	return &tagsController{
+		ec2:   ec2,
+		elbv2: elbv2,
+	}
+}
+
+type tagsController struct {
+	ec2   ec2iface.EC2API
+	elbv2 elbv2iface.ELBV2API
+}
+
+func (c *tagsController) Reconcile(ctx context.Context, desired *Tags) error {
+	if strings.HasPrefix(desired.Arn, "arn:aws:elasticloadbalancing") {
+		return c.reconcileElasticloadbalancing(ctx, desired)
+	}
+	return fmt.Errorf("%v tags not implemented", desired.Arn)
+}
+
+func (c *tagsController) reconcileElasticloadbalancing(ctx context.Context, desired *Tags) error {
+	current := NewTags()
+
+	resp, err := c.elbv2.DescribeTags(&elbv2.DescribeTagsInput{ResourceArns: []*string{aws.String(desired.Arn)}})
+	if err != nil {
+		return err
+	}
+
+	for _, tagDescription := range resp.TagDescriptions {
+		if aws.StringValue(tagDescription.ResourceArn) != desired.Arn {
+			continue
+		}
+		for _, tag := range tagDescription.Tags {
+			current.Tags[aws.StringValue(tag.Key)] = aws.StringValue(tag.Value)
+		}
+	}
+
+	modify, remove := tagsChangeSet(current, desired)
+
+	if len(modify) > 0 {
+		albctx.GetLogger(ctx).Infof("Modifying tags on %v to %v", desired.Arn, log.Prettify(modify))
+
+		addParams := &elbv2.AddTagsInput{
+			ResourceArns: []*string{aws.String(desired.Arn)},
+			Tags:         mapToELBV2(modify),
+		}
+		if _, err := c.elbv2.AddTags(addParams); err != nil {
+			if eventf, ok := albctx.GetEventf(ctx); ok {
+				eventf(api.EventTypeWarning, "ERROR", "Error tagging %s: %s", desired.Arn, err.Error())
+			}
+			return err
+		}
+	}
+
+	if len(remove) > 0 {
+		albctx.GetLogger(ctx).Infof("Removing %v tags from %v", strings.Join(remove, ", "), desired.Arn)
+
+		removeParams := &elbv2.RemoveTagsInput{
+			ResourceArns: []*string{aws.String(desired.Arn)},
+			TagKeys:      aws.StringSlice(remove),
+		}
+
+		if _, err := c.elbv2.RemoveTags(removeParams); err != nil {
+			if eventf, ok := albctx.GetEventf(ctx); ok {
+				eventf(api.EventTypeWarning, "ERROR", "Error tagging %s: %s", desired.Arn, err.Error())
+			}
+			return err
+		}
+	}
+
+	return nil
+}
+
+// tagsChangeSet compares b to a, returning a map of tags to add/change to a and a list of tags to remove from a
+func tagsChangeSet(a, b *Tags) (map[string]string, []string) {
+	modify := make(map[string]string)
```

Comment:
this aligns with current behavior iirc.

But should we be more restrictive.
e.g. only maintain tags begin with specific prefix like alb.k8s.io. (I believe customers want to add their own custom tags too).
Similar situation can apply to securityGroup rules, i'm considering use some fixed prefix as description, and only manage rules with such prefix, which will allow customers add their own custom rules

Reply:
as long as we support custom tags via annotation, this should be fine for now 😄